### ISSUE TITLE
Weird failure

### ISSUE THREAD

**Issue by zoffixznet** [ugexe/zef#87]
I got his weird install failure: 

```
zoffix@zoffix-VirtualBox:/tmp/tmp.iDjXwotTqf$ zef --/test install Bailador
===> Searching for: Bailador
===> Found: Bailador
===> Dependencies: HTTP::Easy, Template::Mojo, Template::Mustache, URI, Digest::HMAC, Digest
===> Filtering: Bailador
===> Filtering [OK] for Bailador
===> # SKIP: No Build.pm for Bailador
===> Installing: Bailador
===> Install [FAIL] for Bailador: Use of uninitialized value <element> of type Any in string context
Any of .^name, .perl, .gist, or .say can stringify undefined things, if needed.  in block  at /home/zoffix/.rakudobrew/moar-nom/install/share/perl6/site/sources/E123DDDDFFF94BC860AC2B9C74AAD0EDC91CB4E0 line 5
===SORRY!=== Error while compiling /tmp/tmp.iDjXwotTqf/site#sources/E123DDDDFFF94BC860AC2B9C74AAD0EDC91CB4E0
An exception occurred while evaluating a constant
at /tmp/tmp.iDjXwotTqf/site#sources/E123DDDDFFF94BC860AC2B9C74AAD0EDC91CB4E0:5
Exception details:
  Invalid mime.types file: /home/zoffix/.rakudobrew/moar-nom/install/share/perl6/site/resources
    in method new at /home/zoffix/.rakudobrew/moar-nom/install/share/perl6/site/sources/22FBA36C8B8920A36EC51E9D37131C69CF7809BB (MIME::Types) line 10
    in block  at /home/zoffix/.rakudobrew/moar-nom/install/share/perl6/site/sources/E123DDDDFFF94BC860AC2B9C74AAD0EDC91CB4E0 line 5


Use of uninitialized value <element> of type Any in string context
Any of .^name, .perl, .gist, or .say can stringify undefined things, if needed.  in block  at /home/zoffix/.rakudobrew/moar-nom/install/share/perl6/site/sources/E123DDDDFFF94BC860AC2B9C74AAD0EDC91CB4E0 line 5
===SORRY!=== Error while compiling /tmp/tmp.iDjXwotTqf/site#sources/E123DDDDFFF94BC860AC2B9C74AAD0EDC91CB4E0
An exception occurred while evaluating a constant
at /tmp/tmp.iDjXwotTqf/site#sources/E123DDDDFFF94BC860AC2B9C74AAD0EDC91CB4E0:5
Exception details:
  Invalid mime.types file: /home/zoffix/.rakudobrew/moar-nom/install/share/perl6/site/resources
    in method new at /home/zoffix/.rakudobrew/moar-nom/install/share/perl6/site/sources/22FBA36C8B8920A36EC51E9D37131C69CF7809BB (MIME::Types) line 10
    in block  at /home/zoffix/.rakudobrew/moar-nom/install/share/perl6/site/sources/E123DDDDFFF94BC860AC2B9C74AAD0EDC91CB4E0 line 5


  in block  at /home/zoffix/.rakudobrew/moar-nom/install/share/perl6/site/sources/1DC0BAA246D0774E7EB4F5119C6168E0D8266EFA (Zef::Client) line 479
  in code  at /home/zoffix/.rakudobrew/moar-nom/install/share/perl6/site/sources/1DC0BAA246D0774E7EB4F5119C6168E0D8266EFA (Zef::Client) line 452
  in method install at /home/zoffix/.rakudobrew/moar-nom/install/share/perl6/site/sources/1DC0BAA246D0774E7EB4F5119C6168E0D8266EFA (Zef::Client) line 443
  in sub MAIN at /home/zoffix/.rakudobrew/moar-nom/install/share/perl6/site/sources/A9948E7371E0EB9AFDF1EEEB07B52A1B75537C31 (Zef::CLI) line 150
  in block <unit> at /home/zoffix/.rakudobrew/moar-nom/install/share/perl6/site/resources/21D746CFA68C6190C43E2F84F56A4DF747BA50FA line 1

zoffix@zoffix-VirtualBox:/tmp/tmp.iDjXwotTqf$ 
```

However, the module actually *did* install. But, when I try to launch my app, I get:

```
===SORRY!===
Missing or wrong version of dependency 'site#sources/06F75D046213CAD92B84F66E5F161C084878D0C3' (from 'site#sources/A1907C1BED80809C6FC16BB8693FE6BCD5769B7D (Bailador::Plugin::AssetPack::SASS)')
```

The module mentioned in the error does `depends` on Bailador, but it does not specify any versions. It was installed and my app was working fine before I did these things, some of which may be the cause of the issue:

* Uninstall already-installed `Bailador` / install it from local directory that contains modified version
* Repeat the above step several times
* Uninstall bailador and then just run `zef --/test install Bailador` to get the "normal" version from the ecosystem

And the `mime.types` resource mentioned in the install failure error seems to be this one, from my https://github.com/zoffixznet/perl6-Bailador-Plugin-Static that does have `Bailador` as prereq.

So, it seems like reinstalling a module that is a dependency of some modules causes those dependencies to fail when used.

**Comment by ugexe** [comment]
I suspect if you run with `--ll-exception` the trace will lead to `CompUnit::Repository::Installation.install`. NQP errors that happen during the install phase are also indicative of `CompUnit::Repository::Installation.install`. An additional clue is that the module is actually installed, and fails when precompiling.

This leads to the question: is this a zef or rakudo bug? On one hand zef could reinstall all reverse dependencies. On the other hand if precompilation was disabled this problem would not occur, and because all precomp happens automatically it puts the onus to a correct solution on `CompUnit::Repository::Installation`.

Personally I don't consider reinstalling all reverse dependencies to be anything more than a hacky workaround; The correct solution is to re-precompile the reverse dependencies, but the only way for a package manager to do that is to do a complete reinstall or to nuke entire precomp directories. `CUR` should really be handling that automatically.

tl;dr: rakudo is failing to invalidate precomp files in situations that it actually should

**Comment by ugexe** [comment]
This may be a good reason to finally design the hook system for the various phases. This will allow a workaround (as well as the current Build.pm cruft) to act as compatibility adapters alongside similar before/after style phase modifiers. It will also make it easier to drop (or at least bury) the technical debt of future "temporary" solutions. The current plugin system can probably be adapted to take on this role, essentially allow hooks to `wrap` each phase plugin (each hook is guaranteed to be called whereas plugins act similar to a lazy iterator where the last plugin in the chain may never need to be called)

In such a system I think the solution is then a `pre-install` hook script that will delete precomp files for reverse dependencies if the rakudo version is less than $whenever-its-fixed

**Comment by FROGGS** [comment]
Please test again with latest rakudo. This patch might improve the situation: https://github.com/rakudo/rakudo/commit/d0a00164e9

**Comment by zoffixznet** [comment]
Can't reproduce any longer. Thanks!
